libflowlimit/QPSLimiter.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * @CopyRight:
+ * FISCO-BCOS is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * FISCO-BCOS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with FISCO-BCOS.  If not, see <http://www.gnu.org/licenses/>
+ * (c) 2016-2020 fisco-dev contributors.
+ */
+/**
+ * @brief : Implement of QPSLimiter
+ * @file: QPSLimiter.cpp
+ * @author: yujiechen
+ * @date: 2020-04-15
+ */
+#include "QPSLimiter.h"
+#include <thread>
+
+using namespace dev;
+using namespace dev::limit;
+
+QPSLimiter::QPSLimiter(uint64_t const& _maxQPS) : m_maxQPS(_maxQPS)
+{
+    m_permitsUpdateInterval = (double)1000000 / (double)m_maxQPS;
+    m_lastPermitsUpdateTime = utcSteadyTimeUs();
+    QPSLIMIT_LOG(DEBUG) << LOG_KV("permitsUpdateInterval", m_permitsUpdateInterval);
+}
+
+void QPSLimiter::acquire(int64_t const& _requiredPermits)
+{
+    int64_t waitTime = fetchPermitsAndGetWaitTime(_requiredPermits);
+    // sleep
+    if (waitTime > 0)
+    {
+        std::this_thread::sleep_for(std::chrono::microseconds(waitTime));

```

Comment:
the reason this sleep can't be deleted is?

Reply:
This is used to handle the request smoothly.